lib/client.js

```diff
@@ -298,7 +333,9 @@ class Client extends DispatcherBase {
   [kDispatch] (opts, handler) {
     const origin = opts.origin || this[kUrl].origin
 
-    const request = new Request(origin, opts, handler)
+    let request
+    if (this[kHTTP2Session]) request = Request[kHTTP2BuildRequest](origin, opts, handler)
```

Comment:
I've tried to handle the headers fact in a way that can be decoupled from the original H1 implementation, but the implementation is quite tight and the refactor will require to touch several pieces.

In a best effort, will assume at the first instance that a request is made through H1 building the request as usual.
As at this point, we don't know yet the ALPN protocol, we'll convert to H2 headers only once we already negotiated this with the server, where we'll convert them from string to object one time only.
After the initial check, the rest of the Request through the Client will go as H2 requests, using the `headers` as an object.

Let me know if I missed something or if there are improvements that can be done

cc: @ronag @mcollina

Reply:
sgtm